File: packages/typescript-sdk/src/composition/retry.ts
/**
 * Retry policy for {@link withRetry}.
 */
export type RetryOptions<TError = unknown> = {
  /** Number of total attempts including the first call. */
  readonly attempts?: number;
  /** Initial retry delay in milliseconds. */
  readonly delayMs?: number;
  /** Exponential backoff multiplier applied after each failed attempt. */
  readonly factor?: number;
  /** Upper bound for backoff delay in milliseconds. */
  readonly maxDelayMs?: number;
  /**
   * Jitter strategy applied to retry delays.
   *
   * - `"full"` — randomizes delay between 0 and the computed backoff value
   *   to prevent thundering herd when many clients retry simultaneously.
   * - `"none"` — uses the exact computed backoff value with no randomization.
   *
   * @default "full"
   */
  readonly jitter?: "full" | "none";
  /** Predicate to decide whether a failure should be retried. */
  readonly shouldRetry?: (
    error: TError,
    context: { readonly attempt: number; readonly maxAttempts: number }
  ) => boolean;
  /** Optional AbortSignal to cancel retries. */
  readonly signal?: AbortSignal;
};

const wait = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

const checkAborted = (signal?: AbortSignal): void => {
  if (signal?.aborted) {
    throw (signal.reason as Error) ?? new Error("retry aborted");
  }
};

const shouldRetryError = <TError>(
  error: unknown,
  attempt: number,
  maxAttempts: number,
  shouldRetry?: (
    error: TError,
    context: { readonly attempt: number; readonly maxAttempts: number }
  ) => boolean
): boolean => {
  if (attempt >= maxAttempts) {
    return false;
  }
  return shouldRetry?.(error as TError, { attempt, maxAttempts }) ?? true;
};

const computeDelay = (baseDelay: number, jitter: "full" | "none"): number =>
  jitter === "full" ? Math.round(Math.random() * baseDelay) : baseDelay;

/**
 * Executes an async operation with exponential backoff retries.
 *
 * Throws the last observed error when retries are exhausted or when
 * `shouldRetry` returns `false`.
 */
export const withRetry = async <TOutput, TError = unknown>(
  operation: () => Promise<TOutput>,
  options?: RetryOptions<TError>
): Promise<TOutput> => {
  const maxAttempts = Math.max(1, options?.attempts ?? 3);
  const factor = options?.factor ?? 2;
  const maxDelayMs = options?.maxDelayMs ?? 30_000;
  const jitter = options?.jitter ?? "full";

  let attempt = 0;
  let delayMs = options?.delayMs ?? 250;

  for (;;) {
    attempt += 1;
    checkAborted(options?.signal);

    try {
      return await operation();
    } catch (error) {
      if (
        !shouldRetryError(error, attempt, maxAttempts, options?.shouldRetry)
      ) {
        throw error;
      }

      await wait(computeDelay(delayMs, jitter));
      delayMs = Math.min(maxDelayMs, Math.max(1, Math.round(delayMs * factor)));
    }
  }
};

File: packages/go-sdk/composition/retry.go
package composition

import (
	"context"
	"math"
	"math/rand/v2"
	"time"
)

// JitterStrategy controls how retry delays are randomized.
type JitterStrategy string

const (
	JitterFull JitterStrategy = "full"
	JitterNone JitterStrategy = "none"
)

// RetryOptions configures retry behavior.
type RetryOptions struct {
	// Attempts is the total number of attempts including the first call.
	Attempts int
	// DelayMs is the initial retry delay in milliseconds.
	DelayMs int
	// Factor is the exponential backoff multiplier.
	Factor float64
	// MaxDelayMs is the upper bound for backoff delay.
	MaxDelayMs int
	// Jitter controls delay randomization. Default is "full".
	Jitter JitterStrategy
	// ShouldRetry decides whether a failure should be retried.
	ShouldRetry func(err error, attempt int, maxAttempts int) bool
}

func (o *RetryOptions) defaults() {
	if o.Attempts <= 0 {
		o.Attempts = 3
	}
	if o.DelayMs <= 0 {
		o.DelayMs = 250
	}
	if o.Factor <= 0 {
		o.Factor = 2
	}
	if o.MaxDelayMs <= 0 {
		o.MaxDelayMs = 30000
	}
	if o.Jitter == "" {
		o.Jitter = JitterFull
	}
}

func computeDelay(baseDelay int, jitter JitterStrategy) int {
	if jitter == JitterFull {
		return int(math.Round(rand.Float64() * float64(baseDelay)))
	}
	return baseDelay
}

// WithRetry executes an operation with exponential backoff retries.
// It respects context cancellation.
func WithRetry[T any](ctx context.Context, fn func() (T, error), opts *RetryOptions) (T, error) {
	var o RetryOptions
	if opts != nil {
		o = *opts
	}
	o.defaults()

	var zero T
	delayMs := o.DelayMs

	for attempt := 1; ; attempt++ {
		if err := ctx.Err(); err != nil {
			return zero, err
		}

		result, err := fn()
		if err == nil {
			return result, nil
		}

		if attempt >= o.Attempts {
			return zero, err
		}

		if o.ShouldRetry != nil && !o.ShouldRetry(err, attempt, o.Attempts) {
			return zero, err
		}

		waitMs := computeDelay(delayMs, o.Jitter)
		select {
		case <-ctx.Done():
			return zero, ctx.Err()
		case <-time.After(time.Duration(waitMs) * time.Millisecond):
		}

		delayMs = int(math.Min(float64(o.MaxDelayMs), math.Max(1, math.Round(float64(delayMs)*o.Factor))))
	}
}
